### ISSUE TITLE
function_ref.h implementation uses a feature that's removed in C++20

### ISSUE THREAD

**Issue by maxgolov** [open-telemetry/opentelemetry-cpp#104]
This code would no longer compile in strict mode with C++20 enabled in Visual Studio 2019 (16.6.2, tools v142):

https://github.com/open-telemetry/opentelemetry-cpp/blob/9e5924c523c3cd36ecbc38be1c93b6bb162a84d3/api/include/opentelemetry/nostd/function_ref.h#L66

**std::result_of** has been permanently removed in C++20. Reference:
https://en.cppreference.com/w/cpp/types/result_of

Implementation needs to be rewritten.

One possible approach is to backport std::result_of into nostd::result_of 

The other implementation approach is to rework this code to completely avoid the construct.

Visual Studio 2019 compiler takes very strict stance on removed features. There's also a temporary workaround flag that allows to enable deprecated features ( **_HAS_DEPRECATED_RESULT_OF** ). But I believe it is incorrect for common SDK API header to rely on a feature that's been permanently removed from C++ standard.

I also tested this to confirm that recent gcc-9 and clang-10 are both OK with that deprecated construct, even if a module is compiled in C++20 mode.. But what Visual C++ is doing (not supporting removed features) - seems like the right thing to do.

**Comment by maxgolov** [comment]
Possible fix (pseudocode):

```cpp
#if (__cplusplus >= 201703L) || (_STL_LANG > 201703L)
          std::is_convertible<typename std::invoke_result<F, Args...>::type, R>::value,
#else
          std::is_convertible<typename std::result_of<F &(Args...)>::type, R>::value,
#endif
```

I can take a look at it. I hit it when I've been experimenting to measure perf of STL vs nostd, plus verifying ABI compat with different compiler standard flags.

**Comment by maxgolov** [comment]
Fixed by #115
